src/projection.jl

```diff
@@ -0,0 +1,55 @@
+using LinearAlgebra: Diagonal, diag
+
+"""
+    project(T::Type, x, dx)
```

Comment:
Should we make `T` optional to avoid duplication when it is the same as `typeof(x)`?

Reply:
Hmm yeah -- we'd need to privelege a particular representation (presumably `Tangent`), but I think that this would make sense.

Reply:
Is it necessary to close over `x`? I was thinking elsewhere that the right pattern is
```
info = preproject(x)
...
dx = project(dx_raw, info...)
```
with
```
preproject(x) = ()
preproject(x::Number) = (typeof(x),)
preproject(x::AbstractArray) = (typeof(x), axes(x))
```
This lets you pass information besides the type when that's required. (I think the motivating case needed a reshape.)

Reply:
I agree that you could do this, but since we'll be getting opaque closures in 1.7, can we just assume that we'll get to use them, and therefore not have to worry about optimising this stuff away?

Reply:
Oh that's true. What is the status BTW, I haven't kept track, had an idea they existed but were still slow?

Reply:
I think they're in? @oxinabox do you know anything more about the status of opaque closures?

Reply:
I don't think opaque closures are able to fully optimize such cases yet and it's likely not super trivial. In general, I would be careful with relying on potential compiler optimizations in the future for the current design. That's not to say overly optimizing this case is necessarily worth the added API complexity.

Reply:
I agree with everything @simeonschaub said

Reply:
Besides going Tuple -> Matrix (which comes from Zygote's difficulty with splatting arrays, could be solved separately to this) are there other examples where `project` needs `x` not just `typeof(x)`? Passing the type alone is certainly low-API-complexity.

Reply:
Pretty much any `Tangent -> AbstractArray` conversion needs size info I think, you probably need to know about the locations of non-zero elements in sparse arrays, that kind of thing. Off the top of my head, I can't think of others though.

Reply:
Re sparse arrays... is it obvious how many structural zeros `gradient(f, sinpi.(sparse(0:0.5:2)))` should have? Maybe there's an issue discussing this but it doesn't super-obvious to me; perhaps there are more realistic examples than mine which would clarify what should happen. (I don't mean structurally sparse like Diagonal, obviously, I mean SparseArrays.)

Reply:
My claim is that it should have precisely the same number as the primal -- i.e. zeros should be treated structurally (it's what I argue for in the Abstract Arrays Dilemma docs). This is something that we need to reach a consensus on though.

Reply:
OK, but a bigger question than this mini-thread. Are there other clear, concrete, examples?

Reply:
Indeed. Off the top of my head, I can't think of any other obvious ones. Possibly it's going to be limited to array-like things, because they're pretty much the only case in which we need to transform between `Tangent`s and array-like representations of tangents.